core/ledger/pvtdatastorage/retroactive_hashed_index.go

```diff
@@ -0,0 +1,161 @@
+*/
+
+package pvtdatastorage
+
+import (
+	"github.com/hyperledger/fabric-protos-go/ledger/rwset"
+	"github.com/hyperledger/fabric/common/ledger/util/leveldbhelper"
+
+	"github.com/hyperledger/fabric/core/ledger"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/rwsetutil"
+	"github.com/hyperledger/fabric/core/ledger/util"
+	"github.com/pkg/errors"
+)
+
+const (
+	previousDataVersion = ""
+	currentDataVersion  = "2.6"
+	maxUpgradeBatchSize = 4 * 1024 * 1024 // 4 MB
+)
+
+func checkAndConstructHashedIndex(storePath string, ledgerIDs []string) error {
+	i, err := leveldbhelper.RetrieveDataFormatInfo(storePath)
+	if err != nil {
+		return err
+	}
+
+	if i.IsDBEmpty || i.FormatVerison == currentDataVersion {
+		return nil
+	}
+
+	if i.FormatVerison == previousDataVersion {
+		if err := constructHashedIndex(storePath, ledgerIDs); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return errors.Errorf("cannot upgrade data format for pvtdatastore from %s to %s", i.FormatVerison, currentDataVersion)
+}
+
+// constructHashedIndex creates the HashedIndex entries for the private data keys and at the end sets the
+// data format version to the current version (2.6)
+func constructHashedIndex(storePath string, ledgerIDs []string) error {
+	p, err := leveldbhelper.NewProvider(
+		&leveldbhelper.Conf{
+			DBPath:         storePath,
+			ExpectedFormat: previousDataVersion,
+		})
+	if err != nil {
+		return err
+	}
+
+	defer p.Close()
+
+	for _, l := range ledgerIDs {
+		db := p.GetDBHandle(l)
+		if err := constructHashedIndexFor(l, db); err != nil {
+			return err
+		}
+	}
+	return p.SetDataFormat(currentDataVersion)
+}
+
+// constructHashedIndexFor creates the HashedIndex entries for a given ledger.
+// In this function we also piggyback to upgrade any private data key from format V11 to V12
+func constructHashedIndexFor(ledgerID string, db *leveldbhelper.DBHandle) error {
+	startKey, endKey := entireDatakeyRange()
+	itr, err := db.GetIterator(startKey, endKey)
+	if err != nil {
+		return err
+	}
+	defer itr.Release()
+
+	batch := db.NewUpdateBatch()
+
+	for itr.Next() {
+		k := itr.Key()
+		v := itr.Value()
+
+		// In pvtdatastore, we used to persist the data of entire transaction as a single KV in V11.
+		// Later, when we introduced BlockToLive, we started storing each collection data as a KV.
+		// Here we leverage this opportunity (when we create HashedIndexKeys retroactively) to convert
+		// the private data in V11 format, if any so in the next available opportunity,
+		// we will remove the code where we have to detect this difference while processing queries on pvtdatastore.
+		v11Fmt, err := v11Format(k)
+		if err != nil {
+			return err
+		}
+
+		if v11Fmt {
+			blockNum, _, err := v11DecodePK(k)
+			if err != nil {
+				return err
+			}
+
+			txPvtData, err := v11DecodeKV(k, v, nil)
+			if err != nil {
+				return err
+			}
+
+			dataEntries := prepareDataEntries(blockNum, []*ledger.TxPvtData{txPvtData})
```

Comment:
I think you've proven my point by pointing to the struct:

```
type storeEntries struct {
	dataEntries             []*dataEntry
	expiryEntries           []*expiryEntry
	elgMissingDataEntries   map[missingDataKey]*bitset.BitSet
	inelgMissingDataEntries map[missingDataKey]*bitset.BitSet
}
```

Three of the fields have descriptive names which makes them easy to understand. One of the fields has a generic name "dataEntry" which makes it impossible to know what kind of data it is. Everything in the ledger is a "dataEntry". I'm asking to update the name of the type as well as the variable so that readers know what kind of data it is.

Reply:
We are free to draw any conclusion but IMO, it proves other way. `data` means data as `expiry` means expiry, `missingData` mean missing data. I am not sure where the confusion is. I guess that you are applying general meaning to it. Not sure what Reader you have in mind -- These terminologies are there since day one. Anyone who had written code in this package and understood the code/design never raised this issue. Not sure when you say what kind of data - in private store context, of course data means `private data`.  

In any case, changing all over this type and name is beyond the scope of this PR and given that I am not convinced with your observation, we can discuss outside this PR and you are welcome to suggest the change.

Reply:
The only thing I can agree to here is that `dataEntry` preceded this PR and can be discussed outside.
As the code author, `dataEntry` clearly means something very specific to you. As a non-author reader, I can tell you such a general term (everything in the ledger is a data entry) doesn't mean something specific to somebody reading the code for the first time. The definition of private data store data entry has changed over time and even I forget which way was original and which way is current, therefore something like `collDataEntry` would reinforce that the data entries are for private data writes and scoped at collection level.

The fact that it was there since day one is insignificant, I thought we all agreed to review code with a more critical eye towards readability than we have in the past, and improve existing code as we go.

Again, it was a minor point and I'm happy to move on.

Reply:
Yes, I would be happy to agree with the disagreement on this :-)